File: EmulatorGUI/EmulatorLib/RomInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace EmulatorGUI.EmulatorLib
{
	public enum CartridgeType
	{
		ROM_Only = 0x00,
		MBC1,
		MBC1_RAM,
		MBC1_RAM_BATTERY,

		MBC2 = 0x05,
		MBC2_RAM,

		ROM_RAM = 0x08,
		ROM_RAM_BATTERY,

		MMM01 = 0x0B,
		MMM01_RAM,
		MMM01_RAM_BATTERY,

		MBC3_TIMER_BATTERY = 0x0F,
		MBC3_TIMER_RAM_BATTERY,
		MBC3,
		MBC3_RAM,
		MBC3_RAM_BATTERY,

		MBC5 = 0x19,
		MBC5_RAM = 0x1A,
		MBC5_RAM_BATTERY,
		MBC5_RUMBLE,
		MBC5_RUMBLE_RAM,
		MBC5_RUMBLE_RAM_BATTERY,

		MBC6 = 0x20,

		MBC7_SENSOR_RUMBLE_RAM_BATTERY = 0x22,

		POCKET_CAMERA = 0xFC,
		BANDAI_TAMA5 = 0xFD,
		HuC3 = 0xFE,
		HuC1_RAM_BATTERY = 0xFF
	};

    public enum MapperType
	{
		None,
		MBC1,
		MBC2,
		MBC3,
		MBC5,
		MBC6,
		MBC7,
		MMM01,
		HuC1,
		HuC3,
		TAMA5
	};

	public partial class RomInfo : ObjectBase
    {
        [LibraryImport("EmulatorLib.dll")]
        [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
        private static partial nint rominfo_create(byte[] data, int size);

        [LibraryImport("EmulatorLib.dll")]
        [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
        private static partial byte rominfo_getcartridgetype(nint pointer);
        [LibraryImport("EmulatorLib.dll")]
        [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
        private static partial byte rominfo_getmappertype(nint pointer);
        [LibraryImport("EmulatorLib.dll")]
        [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
        private static partial uint rominfo_getramsize(nint pointer);

        [LibraryImport("EmulatorLib.dll")]
        [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
        private static partial byte rominfo_read(nint pointer, uint address);

        public RomInfo(byte[] data) : base(rominfo_create(data, data.Length))
        {

        }


        public string GameName { get; private set; } = string.Empty;

		public byte Revision { get; private set; } = 0;

		public byte RawCartridgeType => (byte)CartridgeType;

		public CartridgeType CartridgeType => (CartridgeType)rominfo_getcartridgetype(CPointer);



        public MapperType MapperType => (MapperType)rominfo_getmappertype(CPointer);
		public bool HasRAM()
		{
            return CartridgeType switch
            {
                CartridgeType.MBC1_RAM or CartridgeType.MBC1_RAM_BATTERY or CartridgeType.MBC2_RAM or CartridgeType.ROM_RAM or CartridgeType.ROM_RAM_BATTERY or CartridgeType.MMM01_RAM or CartridgeType.MMM01_RAM_BATTERY or CartridgeType.MBC3_TIMER_RAM_BATTERY or CartridgeType.MBC3_RAM or CartridgeType.MBC3_RAM_BATTERY or CartridgeType.MBC5_RAM or CartridgeType.MBC5_RAM_BATTERY or CartridgeType.MBC5_RUMBLE_RAM_BATTERY or CartridgeType.MBC7_SENSOR_RUMBLE_RAM_BATTERY or CartridgeType.HuC1_RAM_BATTERY => true,
                _ => false,
            };
        }
		public bool HasBattery()
		{
            return CartridgeType switch
            {
                CartridgeType.MBC1_RAM_BATTERY or CartridgeType.ROM_RAM_BATTERY or CartridgeType.MMM01_RAM_BATTERY or CartridgeType.MBC3_TIMER_BATTERY or CartridgeType.MBC3_TIMER_RAM_BATTERY or CartridgeType.MBC3_RAM_BATTERY or CartridgeType.MBC5_RUMBLE_RAM_BATTERY or CartridgeType.MBC7_SENSOR_RUMBLE_RAM_BATTERY or CartridgeType.HuC1_RAM_BATTERY => true,
                _ => false,
            };
        }
		public bool HasRumble()
		{
            return CartridgeType switch
            {
                CartridgeType.MBC5_RUMBLE or CartridgeType.MBC5_RUMBLE_RAM or CartridgeType.MBC5_RUMBLE_RAM_BATTERY or CartridgeType.MBC7_SENSOR_RUMBLE_RAM_BATTERY => true,
                _ => false,
            };
        }

        public override void Destroy()
        {

        }

        public int ROMSize { get; private set; } = 0;
		public uint RAMSize => rominfo_getramsize(CPointer);

        public int BankCount => ROMSize / 16000;

		public byte Read(uint address) => rominfo_read(CPointer, address);
    }
}

File: EmulatorLib/RomInfo.cpp
#include "RomInfo.hpp"

#include <iostream>

ROMInfo::ROMInfo(uint8_t* rom, uint32_t size)
{
	_rom = new uint8_t[size];
	std::memcpy(_rom, rom, size);

	std::cout << "creating rom with size: " << size << std::endl;
	for (uint16_t i = 0x0134; i < 0x143; i++)
	{
		char c = (char)_rom[i];

		if (c == 0x00)
			break;

		_gameName += c;
	}

	_cartType = _rom[0x147];

	_romSize = 32000 << _rom[0x148];
	switch (_rom[0x148])
	{
	case 0x01:
		_ramSize = 2000;
		break;
	case 0x02:
		_ramSize = 8000;
		break;
	case 0x03:
		_ramSize = 32000;
		break;
	case 0x04:
		_ramSize = 128000;
		break;
	case 0x05:
		_ramSize = 64000;
		break;
	default:
		_ramSize = 0x00;
		break;
	}
	_ramSize = 32000 << _rom[0x148];

	_revision = _rom[0x14C];
}
ROMInfo::~ROMInfo()
{
	if (_rom != nullptr)
		delete _rom;

	_rom = nullptr;
	size = 0;
}

const std::string& ROMInfo::GetGameName() const
{
	return _gameName;
}
uint8_t ROMInfo::GetRevision() const
{
	return _revision;
}

uint8_t ROMInfo::GetRawCartridgeType() const
{
	return _cartType;
}

CartridgeType ROMInfo::GetCartridgeType() const
{
	return (CartridgeType)_cartType;
}
MapperType ROMInfo::GetMapperType() const
{
	switch ((CartridgeType)_cartType)
	{
	case CartridgeType::ROM_Only:
	case CartridgeType::ROM_RAM:
	case CartridgeType::ROM_RAM_BATTERY:
	case CartridgeType::POCKET_CAMERA:
		return MapperType::None;

	case CartridgeType::MBC1:
	case CartridgeType::MBC1_RAM:
	case CartridgeType::MBC1_RAM_BATTERY:
		return MapperType::MBC1;

	case CartridgeType::MBC2:
	case CartridgeType::MBC2_RAM:
		return MapperType::MBC2;

	case CartridgeType::MMM01:
	case CartridgeType::MMM01_RAM:
	case CartridgeType::MMM01_RAM_BATTERY:
		return MapperType::MMM01;

	case CartridgeType::MBC3_TIMER_BATTERY:
	case CartridgeType::MBC3_TIMER_RAM_BATTERY:
	case CartridgeType::MBC3:
	case CartridgeType::MBC3_RAM:
	case CartridgeType::MBC3_RAM_BATTERY:
		return MapperType::MBC3;

	case CartridgeType::MBC5:
	case CartridgeType::MBC5_RAM:
	case CartridgeType::MBC5_RAM_BATTERY:
	case CartridgeType::MBC5_RUMBLE:
	case CartridgeType::MBC5_RUMBLE_RAM:
	case CartridgeType::MBC5_RUMBLE_RAM_BATTERY:
		return MapperType::MBC5;

	case CartridgeType::MBC6:
		return MapperType::MBC6;

	case CartridgeType::MBC7_SENSOR_RUMBLE_RAM_BATTERY:
		return MapperType::MBC7;

	case CartridgeType::BANDAI_TAMA5:
		return MapperType::TAMA5;

	case CartridgeType::HuC1_RAM_BATTERY:
		return MapperType::HuC1;

	case CartridgeType::HuC3:
		return MapperType::HuC3;
	}
}

bool ROMInfo::HasRAM() const
{
	switch ((CartridgeType)_cartType)
	{
	case CartridgeType::ROM_Only:
	case CartridgeType::MBC1:
	case CartridgeType::MBC2:
	case CartridgeType::MMM01:
	case CartridgeType::MBC3_TIMER_BATTERY:
	case CartridgeType::MBC3:
	case CartridgeType::MBC5:	
	case CartridgeType::MBC5_RUMBLE:
	case CartridgeType::MBC5_RUMBLE_RAM:
	case CartridgeType::MBC6:
	case CartridgeType::POCKET_CAMERA:
	case CartridgeType::BANDAI_TAMA5:
	case CartridgeType::HuC3:
		return false;

	case CartridgeType::MBC1_RAM:
	case CartridgeType::MBC1_RAM_BATTERY:
	case CartridgeType::MBC2_RAM:
	case CartridgeType::ROM_RAM:
	case CartridgeType::ROM_RAM_BATTERY:
	case CartridgeType::MMM01_RAM:
	case CartridgeType::MMM01_RAM_BATTERY:
	case CartridgeType::MBC3_TIMER_RAM_BATTERY:
	case CartridgeType::MBC3_RAM:
	case CartridgeType::MBC3_RAM_BATTERY:
	case CartridgeType::MBC5_RAM:
	case CartridgeType::MBC5_RAM_BATTERY:
	case CartridgeType::MBC5_RUMBLE_RAM_BATTERY:
	case CartridgeType::MBC7_SENSOR_RUMBLE_RAM_BATTERY:
	case CartridgeType::HuC1_RAM_BATTERY:
		return true;
	}
}
bool ROMInfo::HasBattery() const
{
	switch ((CartridgeType)_cartType)
	{
	case CartridgeType::ROM_Only:
	case CartridgeType::MBC1:
	case CartridgeType::MBC1_RAM:
	case CartridgeType::MBC2:
	case CartridgeType::MBC2_RAM:
	case CartridgeType::ROM_RAM:
	case CartridgeType::MMM01:
	case CartridgeType::MMM01_RAM:
	case CartridgeType::MBC3:
	case CartridgeType::MBC3_RAM:
	case CartridgeType::MBC5:
	case CartridgeType::MBC5_RAM:
	case CartridgeType::MBC5_RAM_BATTERY:
	case CartridgeType::MBC5_RUMBLE:
	case CartridgeType::MBC5_RUMBLE_RAM:
	case CartridgeType::MBC6:
	case CartridgeType::POCKET_CAMERA:
	case CartridgeType::BANDAI_TAMA5:
	case CartridgeType::HuC3:
		return false;
	case CartridgeType::MBC1_RAM_BATTERY:
	case CartridgeType::ROM_RAM_BATTERY:
	case CartridgeType::MMM01_RAM_BATTERY:
	case CartridgeType::MBC3_TIMER_BATTERY:
	case CartridgeType::MBC3_TIMER_RAM_BATTERY:
	case CartridgeType::MBC3_RAM_BATTERY:
	case CartridgeType::MBC5_RUMBLE_RAM_BATTERY:
	case CartridgeType::MBC7_SENSOR_RUMBLE_RAM_BATTERY:
	case CartridgeType::HuC1_RAM_BATTERY:
		return true;
	}
}
bool ROMInfo::HasRumble() const
{
	switch ((CartridgeType)_cartType)
	{
	case CartridgeType::ROM_Only:
	case CartridgeType::MBC1:
	case CartridgeType::MBC1_RAM:
	case CartridgeType::MBC1_RAM_BATTERY:
	case CartridgeType::MBC2:
	case CartridgeType::MBC2_RAM:
	case CartridgeType::ROM_RAM:
	case CartridgeType::ROM_RAM_BATTERY:
	case CartridgeType::MMM01:
	case CartridgeType::MMM01_RAM:
	case CartridgeType::MMM01_RAM_BATTERY:
	case CartridgeType::MBC3_TIMER_BATTERY:
	case CartridgeType::MBC3_TIMER_RAM_BATTERY:
	case CartridgeType::MBC3:
	case CartridgeType::MBC3_RAM:
	case CartridgeType::MBC3_RAM_BATTERY:
	case CartridgeType::MBC5:
	case CartridgeType::MBC5_RAM:
	case CartridgeType::MBC5_RAM_BATTERY:
	case CartridgeType::MBC6:
	case CartridgeType::POCKET_CAMERA:
	case CartridgeType::BANDAI_TAMA5:
	case CartridgeType::HuC3:
	case CartridgeType::HuC1_RAM_BATTERY:
		return false;
	case CartridgeType::MBC5_RUMBLE:
	case CartridgeType::MBC5_RUMBLE_RAM:
	case CartridgeType::MBC5_RUMBLE_RAM_BATTERY:
	case CartridgeType::MBC7_SENSOR_RUMBLE_RAM_BATTERY:
		return true;
	}
}

uint32_t ROMInfo::GetROMSize() const
{
	return _romSize;
}
uint32_t ROMInfo::GetRAMSize() const
{
	return _ramSize;
}

uint8_t ROMInfo::Read(uint32_t address) const
{
	if (_rom == nullptr || address >= _romSize)
		return 0xFF;

	return _rom[address];
}